Tn/add export functionality

TylerNoblett opened the pull request:
Adds an export feature per the [following features](https://github.com/theparanoids/ashirt-server/issues/734#issuecomment-1532289719)

A few things to note:
- a user can export the evidence if the global flag for exporting is enabled AND if they are an admin (system, operation, or part of a group with admin privileges).
- Re: denormlization, I removed the tag ID and the evidence UUID, but let me know if you'd like me to include the evidence UUID and I'll add that back.
- I am saving the images as JPEG and all the other media as txt, but I'm not sure if the terminal recordings are supposed to be another file type; let me know what they should be saved as!

https://user-images.githubusercontent.com/22867407/236897195-5e98f6c6-f577-4eb0-900c-9c1df3d6cad3.mov

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

TylerNoblett commented on docker-compose.yml:
```diff
@@ -80,6 +80,7 @@ services:
     environment:
       - NODE_ENV=development
       - WEB_BACKEND_ORIGIN=http://backend:3000
+      - ENABLE_EVIDENCE_EXPORT=false
```
Global flag

TylerNoblett commented on frontend/src/pages/operation_show/evidence_list/index.tsx:
```diff
@@ -61,14 +72,62 @@ export default () => {
+    const evidenceCopy = _.cloneDeep(evidence)
+    evidenceCopy.forEach(e => {
+      e.tags.forEach(t => {
+        delete (t as DenormalizedTag).id
+        })
+      delete (e as DenormalizedEvidence).uuid
```
removes the 
- tag ID 
- UUID  
let me know if you DO want UUID included

TylerNoblett commented on frontend/src/pages/operation_show/layout/toolbar/index.tsx:
```diff
@@ -45,6 +47,8 @@ export const Toolbar = (props: {
       ]), [props.operationSlug]
     ))
 
+  const userCanExport = process.env.ENABLE_EVIDENCE_EXPORT && props.userCanExportData
```
allows user to export if exporting is enabled AND if they are an admin (system, operation, or part of a group with admin privileges)

TylerNoblett commented on frontend/webpack.config.js:
```diff
@@ -60,6 +61,11 @@ module.exports = (env, argv) => ({
     }),
+    new webpack.DefinePlugin({
+      'process.env': {
+        'ENABLE_EVIDENCE_EXPORT': process.env.ENABLE_EVIDENCE_EXPORT == 'true',
```
one thought that I just had is that I could include this global flag on the backend and incorporate it into `UserCanExportData`, since most of the global flags are on the backend/makes more intuitive sense. I'll do that tomorrow morning

jrozner commented:
Those changes look good. If C/C++ are the only case where the extension isn't clear that seems fine to just use `.c`. If there are many cases where we don't know the exact extension, let's just not add an extension and simplify it.

One other change, can we remove the extra tag information (eg. color) and just make it an array of the tags themselves (eg. ["tag1", "tag2", "tag3"])?

jrozner commented:
Also, can we change the name from `uuid` to `filename`?

TylerNoblett commented:
Got it! I've
- removed the tag color
- changed uuid to filename
- I also added some extra logic to change the file extension to `.cpp` if it's a c/c++ file and the source file is listed with an extension of c++

jrozner commented:
Sorry, I think there was a misunderstanding with the tags. Can you just extract all the names from the tags and make it an array of strings (of the names) rather than an array of objects.

TylerNoblett commented:
> I think there was a misunderstanding with the tags. Can you just extract all the names from the tags and make it an array of strings (of the names) rather than an array of objects.

oh gotcha - I just reread your original message

JoelAtDeluxe commented on backend/services/operations.go:
```diff
@@ -283,6 +284,18 @@ func ReadOperation(ctx context.Context, db *database.Connection, operationSlug s
+	}
+
+	var userCanExportData bool
+	if os.Getenv("ENABLE_EVIDENCE_EXPORT") == "true" && userIsAdmin {
```
This should probably go in webconfig.go and have a function to retrieve the parsed value.
Also, you could do this:

`userCanExportData := os.Getenv("ENABLE_EVIDENCE_EXPORT") == "true"  && userIsAdmin`

TylerNoblett replied:
ah did not know that existed - thanks for the heads up!

TylerNoblett commented on backend/dtos/dtos.go:
```diff
@@ -69,6 +69,7 @@ type Operation struct {
 	TopContribs       []TopContrib  `json:"topContribs"`
 	EvidenceCount     EvidenceCount `json:"evidenceCount,omitempty"`
 	UserCanViewGroups *bool         `json:"userCanViewGroups,omitempty"`
+	UserCanExportData *bool         `json:"userCanExportData,omitempty"`
```
yeah I just copied what was there (and figured there was a good reason for it), but normal bool is fine with me 😄

TylerNoblett commented on frontend/src/pages/operation_show/evidence_list/index.tsx:
```diff
@@ -61,14 +73,96 @@ export default () => {
+      media.forEach((mb) => {
+        let filename;
+        const fileExtension = mb.sourceFilename?.slice(mb.sourceFilename?.lastIndexOf(".") + 1)
+        console.log("fileExtension", fileExtension)
```
ah no

TylerNoblett replied:
addressed in [832](https://github.com/theparanoids/ashirt-server/compare/tn/joels-feedback-on-832)
